### PR TITLE
QueryPayload decoding and use from query params

### PR DESCRIPTION
**Pull request by umbreak**
This is the second step on addressing the issue #77 .
Further changes will implement the /queries endpoints

### CODE REVIEW

File: modules/service/src/main/scala/ch/epfl/bluebrain/nexus/kg/service/directives/StringUnmarshaller.scala [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+        }
+    }
+
+  /**
+    * String => `A`
+    *
+    * @param f the function to convert from String => `Json`
+    * @return unmarshaller for `A`
+    */
+  implicit final def unmarshaller[A](f: (String => Either[Throwable, Json]))(
```
Comment (bogdanromanx) [inline modules/service/src/main/scala/ch/epfl/bluebrain/nexus/kg/service/directives/StringUnmarshaller.scala]: does this need to be implicit?

File: modules/service/src/main/scala/ch/epfl/bluebrain/nexus/kg/service/routes/RejectionHandling.scala [hunk 2]
```diff
@@ -31,10 +31,10 @@ object RejectionHandling {
...
         case MalformedQueryParamRejection(_, _, Some(e: WrongOrInvalidJson)) =>
           complete(BadRequest -> (e: HttpRejection))
-        case MalformedQueryParamRejection(_, _, Some(e: IllegalFilterFormat)) =>
-          complete(BadRequest -> (e: CommonRejections))
-        case MalformedQueryParamRejection(_, _, Some(e: IllegalOutputFormat)) =>
-          complete(BadRequest -> (e: CommonRejections))
+        case MalformedQueryParamRejection(_, _, Some(err)) =>
+          complete(BadRequest -> (IllegalParam(err.getMessage): CommonRejections))
+        case MalformedQueryParamRejection(_, _, _) =>
+          complete(BadRequest -> (IllegalParam(""): CommonRejections))
```
Comment (bogdanromanx) [inline modules/service/src/main/scala/ch/epfl/bluebrain/nexus/kg/service/routes/RejectionHandling.scala]: empty illegal param?

File: modules/service/src/test/scala/ch/epfl/bluebrain/nexus/kg/service/query/QueryPayloadDecoderSpec.scala [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        q = Some("someText"),
+        resource = QueryResource.Schemas,
+        deprecated = Some(false),
+        published = Some(false),
+        format = JsonLdFormat.Expanded,
+        fields = Set(Field("all")),
+        sort = SortList(List(Sort(s"-${nexusBaseVoc}createdAtTime")))
+      ),
+      queryFilterJson -> QueryPayload(
+        `@context` = Json.obj("some" -> Json.fromString("http://www.w3.org/ns/prov#"),
```
Comment (bogdanromanx) [inline modules/service/src/test/scala/ch/epfl/bluebrain/nexus/kg/service/query/QueryPayloadDecoderSpec.scala]: is the duplicate "some" key intended?

File: modules/service/src/main/scala/ch/epfl/bluebrain/nexus/kg/service/routes/RejectionHandling.scala (hunk 2 above)
Comment (umbreak) [reply modules/service/src/main/scala/ch/epfl/bluebrain/nexus/kg/service/routes/RejectionHandling.scala]: When the MalformedQueryParamRejection does not contain a Throwable information

File: modules/service/src/main/scala/ch/epfl/bluebrain/nexus/kg/service/directives/QueryDirectives.scala [hunk 4]
```diff
@@ -46,117 +47,96 @@ trait QueryDirectives {
...
-  def searchQueryParams(implicit qs: QuerySettings, fs: FilteringSettings, ctxs: Contexts[Future])
-    : Directive[(Pagination, Option[Filter], Option[String], Option[Boolean], Set[String], SortList)] =
+  def paramsToQuery(implicit qs: QuerySettings,
+                    fs: FilteringSettings,
+                    ctxs: Contexts[Future]): Directive[(Pagination, QueryPayload)] =
     context.flatMap { jsonCtx =>
-      paginated & filtered(jsonCtx) & q & deprecated & fields(jsonCtx) & sort(jsonCtx)
+      val queryDec = QueryPayloadDecoder(jsonCtx)
+      import queryDec._
+      filtered.flatMap { filter =>
```
[the remark this answers is not in the document]
Comment (umbreak) [reply modules/service/src/main/scala/ch/epfl/bluebrain/nexus/kg/service/directives/QueryDirectives.scala]: I tried that before. It complained about `withFilter` not being implemented for Directive

**Pull request merged**